### ISSUE TITLE
Webpack 5 Module Federation

### ISSUE THREAD

**Issue by billyjov** [ef4/ember-auto-import#455]
Hi, 
Does anyone got Webpack 5 Module federation with Ember works together? i
tried it so far and I get following errors when it come to integrate with the host app:

```js
wrapper.component.ts:23 error loading ember-landscapes-app: TypeError: Cannot read properties of undefined (reading 'length')
    at includes (should-preserve-symlinks.js:1)
    at module.exports (should-preserve-symlinks.js:1)
    at eval (resolve-package-path.js:17)
    at Object../node_modules/resolve-package-path/lib/resolve-package-path.js (chunk.vendors-node_modules_ember-resolver_index_js.59ad9791bbda2580268b.js:212)
    at __webpack_require__ (remoteEntry.js:44)
    at eval (index.js:1)
    at Object../node_modules/resolve-package-path/index.js (chunk.vendors-node_modules_ember-resolver_index_js.59ad9791bbda2580268b.js:179)
    at __webpack_require__ (remoteEntry.js:44)
    at eval (dependency-version-checker.js:1)
    at Object../node_modules/ember-cli-version-checker/src/dependency-version-checker.js (chunk.vendors-node_modules_ember-resolver_index_js.59ad9791bbda2580268b.js:49)
```

Here is my Webpack config (inside the `ember-cli-build.js`):

```js
webpack: {
        resolve: {
          // https://stackoverflow.com/a/47665710/6432698
          symlinks: false,
          modules: ['node_modules'],
          alias: {
            landscape: path.resolve(__dirname, 'app/'),
          },
          fallback: {
            path: require.resolve('path-browserify'),
            os: false,
            fs: false,
            tls: false,
            net: false,
            zlib: false,
            http: false,
            https: false,
            stream: false,
            process: require.resolve('process/browser'),
          },
        },
        // entry: './app/controllers/landscapes.ts',
        entry: './app/app.js',
        output: {
          filename: 'landscapes.js',
          publicPath: 'http://localhost:4500/assets/',
          uniqueName: 'landscapes',
        },
        optimization: {
          // Only needed to bypass a temporary bug
          // runtimeChunk: false,
        },
        // node: {
        //   global: true,
        // },
        plugins: [
          new ModuleFederationPlugin({
            name: 'bundle',
            library: { type: 'var', name: 'landscapes' },
            filename: 'remoteEntry.js',
            exposes: {
              './web-components': './app/app.js',
            },
            shared: [
              'ember-fetch',
              '@glimmer/component',
              '@glimmer/tracking',
              '@ember/optional-features',
              'ember-bootstrap',
              'ember-data',
              'ember-export-application-global',
              'ember-fetch',
              'ember-resolver',
              'ember-source',
              'ember-svg-jar',
            ],
          }),
          // new HtmlWebpackPlugin({
          //   template: './app/index.html',
          // }),
          // //https://newbedev.com/webpack-bundle-js-uncaught-referenceerror-process-is-not-defined
          new webpack.ProvidePlugin({
            process: 'process/browser',
          }),
        ],
      },
```

The demo repository can be found here : https://github.com/billyjov/ember-webpack-issue . Any Idea what i'm doing wrong ? Thanks

**Comment by billyjov** [comment]
Thanks @ef4 and sorry for my late answer.
 
What I wanted to achieve was to make a complete Ember application visible from the outside.
In principle, I would have thought that certain components in the Ember could be made available (as remote) for a host application with the help of module federation.
I now realize that it would be difficult to achieve with `ember-auto-import`, as it only takes care of third-party dependencies.

>If you're trying to pull code into the ember app, I would expect that to work without any particular trouble. 

Do you mean that, for example, a React application that is activated as a remore app with module federation can be consumed by an Ember app? Because according to your statement it would not work with two ember apps (where one consumes the other)

>ember-auto-import uses a webpack build only for the third-party dependencies and manages linking them into the classic non-webpack build pipeline that builds the app. Whereas embroider builds the whole app and dependencies with webpack.

Will it be in the near future to use Embroider completely for the use case. i.e. to consume a complete application as a remote federated module per runtime? I could read from the documentation from Embroier that a current migration from apps to Embroider could be problematic in some places, that some add-ons would not be compatible as a result?

**Comment by ef4** [comment]
> Do you mean that, for example, a React application that is activated as a remore app with module federation can be consumed by an Ember app?

Yes, as long as you know which modules to import out of the react app and how to pass them a Dom element to render into, you could do it all as a custom modifier in an ember app, like:

```hbs
<div {{render-react-component this.thing }} />
```

> Will it be in the near future to use Embroider completely for the use case

Apps are definitely already going to production with Embroider, if that's what you mean. I don't know of anybody who is experimenting with module federation, but since an Embroider app is just using webpack and webpack supports federation, I think federation should work.

**Comment by billyjov** [comment]
Thanks @ef4. I will give a try with a small ember app + Embroider  and see how this can be exposed to be consume in another app as remote app.

**Comment by miketervela** [comment]
@billyjov Did you ever make progress consuming Ember as a remote app?

**Comment by miketervela** [comment]
@ef4 Do you know if it's currently possible to pull an Ember app into another app as local modules or is that part of the effort for mod fed?  I was playing around with stage 2 of embroider but couldn't resolve the modules in vendor.js.  Keep meaning to go back and look at the custom webpack stage.

**Comment by rchrdschfr** [comment]
Has anyone been able to successfully use an Ember app as a host to pull in federated modules from other apps? I have attempted to use `ModuleFederationPlugin` in the `webpackConfig` plugins, specifying the `remotes` which I want to pull in. However when I hit the `import()` statement, I get an error in the console saying that the module could not be found, which tells me that my Ember host is not registering the remotes I am specifying.
